File: src/SearchInsertPosition.java
/**
 * Created by Administrator on 2015/1/19.
 */
public class SearchInsertPosition {
    public int searchInsert(int[] A, int target){
        int l = 0;
        int r = A.length - 1;
        while(true){
            if(target < A[l]){
                return l;
            }else if(target > A[r]){
                return r + 1;
            }else if(target == A[l]){
                return l;
            }else if(target == A[r]){
                return r;
            }

            // A[l] < target < A[r]
            int mid = (l + r) / 2;
            if(mid == l){//r - l = 1
                return r;
            }
            if(A[mid] == target){
                return mid;
            }else if(target < A[mid]){
                r = mid - 1;
            }else{
                l = mid + 1;
            }


        }
    }
    public static void test(){
        SearchInsertPosition instance = new SearchInsertPosition();
        int[] A = new int[]{1,3,5,6};
        System.out.println(instance.searchInsert(A,5));
        System.out.println(instance.searchInsert(A,2));
        System.out.println(instance.searchInsert(A,7));
        System.out.println(instance.searchInsert(A,0));
    }
}

File: cpp/search_insert_position.cpp
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        int n = nums.size();
        if (n == 0) return 0;
        int l = 0;
        int r = n - 1;
        int m = (l + r) / 2;
        int res = 0;
        while(l <= r){
            m = (l + r) / 2;
            if(nums[m] == target){
                return m;
            }else if(nums[m] < target){
                l = m + 1;
                res = 1;
            }else{
                r = m - 1;
                res = 2;
            }
        }
        if(res == 1){
            return m + 1;
        }else if(res == 2){
            return m;
        }
        return m;
    }
};
